### PR TITLE
Improvement Request: Adding exceptions handling in cli/main.py and api/build.py files 

### PR DESCRIPTION
**Pull request by ADRIDEV2024**
Hello, I´am a backend developer that works mostly with Python as freelance and I want to share my little contribution to the main branch for this repo that I really like how it looks !. 

### **CONTEXT**

I have rewieved your source code in detail and navigated for almost all folders and I have come to the conclussion that  we could add exception handling for build_wheel, since a failure to build a wheel should not stop the entire CLI. What do you think about it?

**MY PROPOSAL:**

1. Catch general errors:

Use try-except in the build_wheel function to catch general errors.
Use concurrent.futures.as_completed to catch thread errors and check for exceptions with future.exception().

2. Catch specific errors in _build_wheel_for_platform:

Catch errors in wheel creation and writing, especially when accessing or writing files.

3. Report errors in a clear way:

Show the affected platform and the error.
Add error traces (traceback.format_exc()) to facilitate debugging.

**MY CHANGES:**

1. I/O errors in dist_path.mkdir:

OSError is caught when the output folder cannot be created.
This prevents permission errors or invalid paths.

2. YAML file loading errors:

If load_wheel_spec_from_yaml fails, an error message is displayed with the file name.

3. Catching errors in build_wheel:

Errors in ThreadPoolExecutor threads are caught with future.exception().
An error message is displayed with the error traceback.format_exc()).

4. Errors specific to the _build_wheel_for_platform function:

File errors (OSError, IOError) are handled.
Unchecked errors are printed so that the user sees the full trace.

5. Security and robustness:

With these additions, the CLI tool will not stop at a single error in wheels creation.
Users will receive full details of errors.

SCREENSHOTS WITH THESE CHANGES ADDED:

code(2)
code(1)

⚠NOTE: PLEASE IMPORT TRACEBACK DEPENDENCY TO WORK WHIT TRACEBACK EXCEPTION

#Backend  #BugFixes  #CLITools

### CODE REVIEW

**Comment by timo-reymann** [comment]
Thanks for your contribution, sounds reasonable to me to have more robust error handling.

> ⚠NOTE: PLEASE IMPORT TRACEBACK DEPENDENCY TO WORK WHIT TRACEBACK EXCEPTION

Not quite sure what you mean by this though.

File: binary_wheel_builder/api/build.py [hunk 1]
```diff
@@ -114,22 +114,31 @@ def build_wheel(wheel_meta: Wheel, dist_folder: Path, worker_count: int = 1) ->
...
-            yield future.result()
-
+            for future in concurrent.futures.as_completed(futures):
+                if future.exception() is not None:
+                    print(f"Error in thread: {future.exception()}\n{traceback.format_exc()}")
+                else:
+                    try:
+                        yield future.result()
+                    except Exception as e:
+                        print(f"Error retrieving result from future: {e}\n{traceback.format_exc()}")
```
Comment (timo-reymann) [inline binary_wheel_builder/api/build.py:124]: chore: Rephrase this a bit more user-friendly, imho it is not relevant to the user that this was an error in a thread

**Review by timo-reymann — CHANGES_REQUESTED** [review]

**Comment by ADRIDEV2024** [comment]
> Thanks for your contribution, sounds reasonable to me to have more robust error handling.
> 
> > ⚠NOTE: PLEASE IMPORT TRACEBACK DEPENDENCY TO WORK WHIT TRACEBACK EXCEPTION
> 
> Not quite sure what you mean by this though.

Hi, I have put that note just to clarify that you need to import the “traceback” library, because in the caption where this library is used, the import of this library does not appear because I forgot to capture it, and maybe you could think that it comes by default in Python. It was only for that reason

File: binary_wheel_builder/api/build.py (hunk 1 above)
Comment (ADRIDEV2024) [reply binary_wheel_builder/api/build.py:124]: Yes, you´re right I have to think what another message to show for a better understanding, and maybe replace print with logging, thanks !

File: binary_wheel_builder/api/build.py [hunk 2]
```diff
@@ -114,22 +114,31 @@ def build_wheel(wheel_meta: Wheel, dist_folder: Path, worker_count: int = 1) ->
...
         )
+        with open(wheel_path, 'rb') as wheel:
+            return WheelPlatformBuildResult(
+                checksum=hashlib.sha256(wheel.read()).hexdigest(),
+                file_path=wheel_path,
+            )
+  except (OSError, IOError) as e:
+        raise RuntimeError(f"File operation failed for platform {python_platform}: {e}")
+  except Exception as e:
+        print(f"Unhandled exception in _build_wheel_for_platform for platform {python_platform}: {e}\n{traceback.format_exc()}")
```
[the remark this answers is not in the document]
Comment (ADRIDEV2024) [reply binary_wheel_builder/api/build.py:143]: Ok, I will consider that

**Pull request closed without merging**